ccm_web/client/src/pages/GradebookCanvas.tsx

```diff
@@ -0,0 +1,390 @@
+  dialog: {
+    textAlign: 'center',
+    maxWidth: '75%',
+    margin: 'auto',
+    marginBottom: 15,
+    paddingLeft: 10,
+    paddingRight: 10,
+    width: '75%',
+    '& ol': {
+      margin: 'auto',
+      width: '75%'
+    },
+    '& li': {
+      textAlign: 'left'
+    }
+  },
+  dialogIcon: {
+    color: 'red'
+  }
+}))
+
+interface GradebookRecord {
+  'Current Grade': string
+  'Final Grade': string
+  'SIS Login ID': string
+  grade: string | undefined
+  Student: string
+}
+
+interface GradebookCanvasPageStateData {
+  state: GradebookCanvasPageState
+  invalidations?: GradebookRowInvalidation[] | undefined
+  errorMessage?: JSX.Element[]
+  grades?: GradebookRecord[]
+}
+
+enum GradebookCanvasPageState {
+  Upload,
+  InvalidUpload,
+  Confirm,
+  Done
+}
+
+interface GradebookRowInvalidation {
+  message: string
+  rowNumber: number
+}
+class CurrentAndFinalGradeMismatchInvalidation implements GradebookRowInvalidation {
+  message: string
+  record: GradebookRecord
+  rowNumber: number
+  constructor (record: GradebookRecord, rowNumber: number) {
+    this.record = record
+    this.rowNumber = rowNumber
+    this.message = 'Current and Final grade mismatch: ' + record.Student + '(' + record['SIS Login ID'] + ')'
+  }
+}
+
+interface GradebookRecordValidator {
+  validate: (record: GradebookRecord, rowNumber: number) => GradebookRowInvalidation[]
+}
+abstract class GradebookValidator implements GradebookRecordValidator {
+  abstract validate: (record: GradebookRecord, rowNumber: number) => GradebookRowInvalidation[]
+}
+class CurrentAndFinalGradeMatchGradebookValidator extends GradebookValidator {
+  validate = (record: GradebookRecord, rowNumber: number): GradebookRowInvalidation[] => {
+    const invalidations: GradebookRowInvalidation[] = []
+    if (record['Final Grade'] !== record['Current Grade']) {
+      invalidations.push(new CurrentAndFinalGradeMismatchInvalidation(record, rowNumber))
+    }
+    return invalidations
+  }
+}
+
+interface DownloadData {
+  data: string
+  fileName: string
+}
+
+function ConvertCanvasGradebook (): JSX.Element {
+  const classes = useStyles()
+  const confirmationClasses = useConfirmationStyles()
+  const rowLevelErrorClasses = useRowLevelErrorStyles()
+  const topLevelClasses = useTopLevelErrorStyles()
+
+  const [pageState, setPageState] = useState<GradebookCanvasPageStateData>({ state: GradebookCanvasPageState.Upload })
+  const [file, setFile] = useState<File|undefined>(undefined)
+  const [downloadData, setDownloadData] = useState<DownloadData|undefined>(undefined)
+
+  const uploadComplete = (file: File): void => {
+    setFile(file)
+  }
+
+  useEffect(() => {
+    parseUpload(file)
+  }, [file])
+
+  const parseUpload = (file: File|undefined): void => {
+    if (file === undefined) {
+      resetPageState()
+      return
+    }
+    // This results in an error on the 2nd "header" row for possible scores
+    parse<GradebookRecord>(file, {
```

Comment:
So, there doesn't appear to be anything validating that the CSV is indeed full of `GradebookRecord`s. When I upload an empty `.csv`, or one with different data, I don't see any errors thrown. Is this by design?